text/0026-serialization.md

```diff
@@ -0,0 +1,391 @@
+    
+    // Handle the message
+    void Handle(in NetworkContext context);
+}
+```
+
+Additionally, each message should have a method on it as follows:
+
+```c#
+[MessageHandler(IMessage.MessageType.MyMessageType)]
+static void Receive(ref FastBufferReader reader, in NetworkContext context);
+```
+
+This static method is responsible for constructing an instance of the received type and handling it. The `[MessageHandler]` attribute statically registers this handler to be responsible for this type via ILPP code generation, just like `[ServerRpc]` and `[ClientRpc] `
+
+Handling will be put in its own method so that it's separate from deserialization, which is important in efficient handling of loopback messages.
+
+The second parameter, `NetworkContext`, contains any data that would be necessary for the handling of the message, including:
+
+- The NetworkManager instance associated with this message
+- The client ID the message came from
+- The network channel the message came in through
+- The received message type (in case two messages share a common deserializer and need to switch on this)
+- The timestamp at which the message was received
+
+More information may be added to `NetworkContext` as necessary.
+
+A sample message might look like this:
+
+```c#
+struct Message : IMessage
+{
+    private float f;
+    private int i;
+    private bool b;
+
+    void GetType()
+    {
+        return IMessage.MessageType.Message;
+    }
+
+    void Serialize(ref FastBufferWriter writer)
+    {
+        writer.WriteValue(this);
+    }
+
+    int GetSizeUpperBound()
+    {
+        return UnsafeUtility.SizeOf<Message>();
+    }
+
+    void Handle(in NetworkContext context)
+    {
+        // Handler code
+    }
+
+    [MessageHandler(IMessage.MessageType.Message)]
+    static void Receive(ref FastBufferReader reader, in NetworkContext context)
+    {
+        Message message;
+        reader.ReadValue(message);
+        message.Handle(context);
+    }
+}
+```
+
+## Serialization and Deserialization
+
+The serialization and deserialization is done via `FastBufferWriter` and `FastBufferReader`. These are similar in interface to `NetworkReader` and `NetworkWriter`, having methods for serializing individual types and methods for serializing packed numbers, but in particular provide a high-performance method called `WriteValue()`/`ReadValue()` (for Writers and Readers, respectively) that can extremely quickly write an entire unmanaged struct to a buffer. There's a trade-off of CPU usage vs bandwidth in using this: Writing individual fields is slower (especially when it includes operations on unaligned memory), but allows the buffer to be filled more efficiently.
+
+Taking an example class:
+
+```C#
+struct Message : IMessage
+{
+    private float f;
+    private bool b;
+    private int i;
+
+    void Serialize(ref FastBufferWriter writer);
+}
+```
+
+Serialize can be implemented in two ways:
```

Comment:
I really like this approach where we can either basically do a memory copy or a packed write based on the message type.